### ISSUE TITLE
java.lang.NoSuchMethodError: No interface method sort

### ISSUE THREAD

**Issue by naowang** [zxing/zxing#1358]
java.lang.NoSuchMethodError: No interface method sort(Ljava/util/Comparator;)V in class Ljava/util/List; or its super classes (declaration of 'java.util.List' appears in /system/framework/core-libart.jar)
        at com.google.zxing.qrcode.detector.FinderPatternFinder.selectBestPatterns(FinderPatternFinder.java:616)
        at com.google.zxing.qrcode.detector.FinderPatternFinder.find(FinderPatternFinder.java:164)
        at com.google.zxing.qrcode.detector.Detector.detect(Detector.java:81)
        at com.google.zxing.qrcode.QRCodeReader.decode(QRCodeReader.java:77)
        at com.google.zxing.MultiFormatReader.decodeInternal(MultiFormatReader.java:173)
        at com.google.zxing.MultiFormatReader.decodeWithState(MultiFormatReader.java:87)
        at org.rydc.smart8things.zxing.DecodeHandler.decode(DecodeHandler.java:86)
        at org.rydc.smart8things.zxing.DecodeHandler.handleMessage(DecodeHandler.java:62)
        at android.os.Handler.dispatchMessage(Handler.java:111)
        at android.os.Looper.loop(Looper.java:207)
        at org.rydc.smart8things.zxing.DecodeThread.run(DecodeThread.java:112)

ArrayList#sort() was added in API level 24 and runtimes below API level 24 don't have that method. Looks like your compileSdkVersion is at 24 so you got the code to compile in the first place.

Use Collections.sort(list, comparator) instead.

**Comment by monsieurtanuki** [comment]
Hi @srowen!
I'm working on [openfoodfacts](https://github.com/openfoodfacts) and I'm still using an API 21 Lollipop smartphone.
We're considering using zxing but while developing I stumble upon the same issue.

Is that OK if I open a PR that includes a possibility for the code to work in API 23 and lower (cf. https://github.com/zxing/zxing/commit/f1683e1f4f670a83c37f6858a7b3db70080e04ae)? Something like
```java
if (android.os.Build.VERSION.SDK_INT < 24) {
    Collections.sort(possibleCenters, moduleComparator);
} else {
    possibleCenters.sort(moduleComparator);
}
```

**Comment by srowen** [comment]
I don't think that helps. The byte code still references the method on loading

**Comment by monsieurtanuki** [comment]
I don't share your pessimism but I cannot swear I'm right either.
The bug appears at runtime, looking for a method that does not exist. The `if` would solve that.
I'll send a PR and test today or tomorrow.

**Comment by srowen** [comment]
Try it, sure. Maybe it works

**Comment by monsieurtanuki** [comment]
Hi @srowen! I've just PR'ed #1595.
For the record In our android project we're currently using `com.google.zxing:core:3.5.1`.

**Comment by thestinger** [comment]
@srowen You can definitely use a conditional check for API level. It's a very common Android development pattern. It only hits an exception when actually trying to use the non-existent method rather than at load time.

**Comment by srowen** [comment]
YOu'd have to use reflection for that, I'm pretty sure, or fails when loading the bytecode

**Comment by thestinger** [comment]
It doesn't fail on Android, at least. It's a very common pattern to gate using newly introduced methods based on API level that way. You can see from the traceback above that they're getting a dynamic NoSuchMethodError.

**Comment by monsieurtanuki** [comment]
@thestinger It's already merged in #1595, and anyway here it's more java than android.
